### PR TITLE
Add springboot-category-model

### PR DESCRIPTION
**Pull request by AntonZhdanov**

### CODE REVIEW

File: src/main/java/com/example/bookstore/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        return categoryService.save(categoryDto);
+    }
+
+    @Operation(summary = "Get all categories with pagination")
+    @GetMapping
+    public List<CategoryDto> getAll(Pageable pageable) {
+        return categoryService.findAll(pageable);
+    }
+
+    @Operation(summary = "Get a book by ID")
```
Comment (Someboty) [inline src/main/java/com/example/bookstore/controller/CategoryController.java]: Copy+paste – evil)

File: src/main/java/com/example/bookstore/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @Operation(summary = "Update a category by ID")
+    @PutMapping("/{id}")
+    public CategoryDto updateCategory(@PathVariable Long id,
+                                      @RequestBody @Valid CategoryDto categoryDto) {
+        Category category = categoryMapper.toModel(categoryDto);
+        category.setId(id);
+        categoryService.update(category);
+        return categoryMapper.toDto(category);
```
Comment (Someboty) [inline src/main/java/com/example/bookstore/controller/CategoryController.java]: Move it to service

File: src/main/java/com/example/bookstore/controller/CategoryController.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @Operation(summary = "Delete a category by ID")
+    @DeleteMapping("/{id}")
+    public void deleteCategory(@PathVariable Long id) {
+        categoryService.deleteById(id);
+    }
+
+    @Operation(summary = "Search books by category id")
+    @GetMapping("/{id}/books")
+    public List<BookDtoWithoutCategoryIds> getBooksByCategoryId(@PathVariable Long id) {
```
Comment (Someboty) [inline src/main/java/com/example/bookstore/controller/CategoryController.java:67]: You need pageable here too

File: src/main/java/com/example/bookstore/dto/book/BookDtoWithoutCategoryIds.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    @NotEmpty
+    private String title;
+    @NotEmpty
+    private String author;
+    @NotNull
+    @Pattern(regexp = "^(978|979)-\\d{1,5}-\\d{2,7}-\\d{1,6}-[\\dX]$")
+    private String isbn;
+    @NotNull
+    @Min(0)
+    @Positive
```
Comment (Someboty) [inline src/main/java/com/example/bookstore/dto/book/BookDtoWithoutCategoryIds.java]: Isn't it do same thing?

File: src/main/java/com/example/bookstore/model/Category.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false)
+    private String name;
+    private String description;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @ManyToMany(mappedBy = "categories")
+    private Set<Book> bookSet = new HashSet<>();
```
Comment (Someboty) [inline src/main/java/com/example/bookstore/model/Category.java]: Why do you need it here?

File: src/main/java/com/example/bookstore/model/Category.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    @Column(nullable = false)
+    private boolean isDeleted = false;
+
+    public void addBook(Book book) {
+        bookSet.add(book);
+        book.getCategories().add(this);
+    }
+}
```
Comment (Someboty) [inline src/main/java/com/example/bookstore/model/Category.java]: Where it used?

File: src/main/java/com/example/bookstore/service/category/CategoryServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        categoryRepository.deleteById(id);
+    }
+
+    @Override
+    public List<BookDtoWithoutCategoryIds> getBooksByCategoriesId(Long id) {
+        Category category = categoryRepository.findById(id).orElseThrow(()
+                -> new EntityNotFoundException("Can't find category by id: " + id));
+
+        Set<Book> books = category.getBookSet();
+        List<BookDtoWithoutCategoryIds> bookDtoWithoutCategoryIdsList = books
```
Comment (Someboty) [inline src/main/java/com/example/bookstore/service/category/CategoryServiceImpl.java]: Maybe add method to repository?

File: src/main/resources/db/changelog/changes/06-create-category-table.yaml [hunk 8]
```diff
@@ -0,0 +1,29 @@
+databaseChangeLog:
+  - changeSet:
+      id: create-category-table
+      author: root
```
Comment (Someboty) [inline src/main/resources/db/changelog/changes/06-create-category-table.yaml]: You have a pretty name)

File: src/main/resources/db/changelog/changes/07-create-books-categories-table.yaml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+                  name: category_id
+                  type: bigint
+                  constraints:
+                    nullable: false
+        - addForeignKeyConstraint:
+            baseTableName: books_categories
+            baseColumnNames: book_id
+            referencedTableName: books
+            referencedColumnNames: id
+            constraintName: fk_book_category_book
```
Comment (Someboty) [inline src/main/resources/db/changelog/changes/07-create-books-categories-table.yaml:34]:
```suggestion
                    foreignKeyName: categories.id
```

**Review by Someboty — CHANGES_REQUESTED** [review]
Great work! But did you check for lazy exception?

File: src/main/java/com/example/bookstore/dto/book/BookDtoWithoutCategoryIds.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    private Long id;
+    @NotEmpty
+    private String title;
+    @NotEmpty
+    private String author;
+    @NotNull
+    @ISBN
+    private String isbn;
+    @NotNull
+    @Min(0)
```
Comment (vovasalyha) [inline src/main/java/com/example/bookstore/dto/book/BookDtoWithoutCategoryIds.java]:
```suggestion
```

File: src/main/java/com/example/bookstore/dto/book/CreateBookRequestDto.java [hunk 11]
```diff
@@ -3,27 +3,18 @@
...
-    private String description;
-    @NotEmpty
-    private String coverImage;
-}
+public record CreateBookRequestDto(
+        Long id,
+        @NotEmpty String title,
+        @NotEmpty String author,
+        @NotNull @ISBN String isbn,
+        @NotNull @Min(0) @Positive BigDecimal price,
```
Comment (vovasalyha) [inline src/main/java/com/example/bookstore/dto/book/CreateBookRequestDto.java]:
```suggestion
        @NotNull @Positive BigDecimal price,
```

File: src/main/java/com/example/bookstore/model/Category.java (hunk 5 above)
Comment (vovasalyha) [reply src/main/java/com/example/bookstore/model/Category.java]: I agree that it is redundant complexity and you can remove it, but it is for you to decide
```suggestion
    private Set<Book> books = new HashSet<>();
```

File: src/main/java/com/example/bookstore/service/category/CategoryServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+
+    @Override
+    public CategoryDto save(CategoryDto categoryDto) {
+        Category category = categoryMapper.toModel(categoryDto);
+        return categoryMapper.toDto(categoryRepository.save(category));
+    }
+
+    @Override
+    public CategoryDto update(Long id, CategoryDto categoryDto) {
+        Category category = categoryMapper.toModel(categoryDto);
```
Comment (vovasalyha) [inline src/main/java/com/example/bookstore/service/category/CategoryServiceImpl.java]: You don't need this and next line

File: src/main/java/com/example/bookstore/service/category/CategoryServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+        return categoryMapper.toDto(category);
+    }
+
+    @Override
+    public CategoryDto save(CategoryDto categoryDto) {
+        Category category = categoryMapper.toModel(categoryDto);
+        return categoryMapper.toDto(categoryRepository.save(category));
+    }
+
+    @Override
```
Comment (vovasalyha) [inline src/main/java/com/example/bookstore/service/category/CategoryServiceImpl.java:44]:
```suggestion
    @Override
    @Transactional
```

File: src/main/java/com/example/bookstore/service/category/CategoryServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+    public void deleteById(Long id) {
+        categoryRepository.deleteById(id);
+    }
+
+    @Override
+    public List<BookDtoWithoutCategoryIds> getBooksByCategoriesId(Long id) {
+        Category category = categoryRepository.findCategoryWithBooks(id);
```
Comment (vovasalyha) [inline src/main/java/com/example/bookstore/service/category/CategoryServiceImpl.java]: Everything can be fetched in one query. Additionally, this method more suits to the services related to the book, not to the category

File: src/main/resources/db/changelog/changes/07-create-books-categories-table.yaml [hunk 15]
```diff
@@ -0,0 +1,30 @@
+databaseChangeLog:
+  - changeSet:
+      id: create-book-category-table
+      author: anton
+      changes:
+        - createTable:
+            tableName: books_categories
+            columns:
```
Comment (vovasalyha) [inline src/main/resources/db/changelog/changes/07-create-books-categories-table.yaml:8]: add also constraint primary key (book_id, category_id)

**Review by vovasalyha — CHANGES_REQUESTED** [review]
GJ. Let's improve it a bit

File: src/main/java/com/example/bookstore/security/JwtAuthenticationFilter.java [hunk 16]
```diff
@@ -18,6 +18,8 @@
 @Component
 @RequiredArgsConstructor
 public class JwtAuthenticationFilter extends OncePerRequestFilter {
+    private static final String AUTHORIZATION_HEADER = "Authorization";
```
Comment (vovasalyha) [inline src/main/java/com/example/bookstore/security/JwtAuthenticationFilter.java]: Spring's HttpHeaders class already has this constant, you can also consider using it

File: src/main/java/com/example/bookstore/service/book/BookServiceImpl.java [hunk 17]
```diff
@@ -66,4 +68,12 @@ public List<BookDto> search(BookSearchParametersDto searchParameters) {
                 .map(bookMapper::toDto)
                 .toList();
     }
+
+    @Override
+    public List<BookDtoWithoutCategoryIds> getBooksByCategoryId(Long categoryId) {
+        List<Book> books = bookRepository.findAllByCategoryId(categoryId);
```
Comment (vovasalyha) [inline src/main/java/com/example/bookstore/service/book/BookServiceImpl.java:73]: can be inlined

**Review by vovasalyha — APPROVED** [review]

**Pull request merged**